frame/dvm/evm/precompiles/state-storage/src/lib.rs

```diff
@@ -0,0 +1,95 @@
+// This file is part of Darwinia.
+//
+// Copyright (C) 2018-2022 Darwinia Network
+// SPDX-License-Identifier: GPL-3.0
+//
+// Darwinia is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Darwinia is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Darwinia. If not, see <https://www.gnu.org/licenses/>.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+#[cfg(test)]
+mod tests;
+
+// --- core ---
+use core::marker::PhantomData;
+// --- crates.io ---
+use ethabi::{ParamType, StateMutability, Token};
+use evm::ExitRevert;
+// --- darwinia-network ---
+use darwinia_evm_precompile_utils::PrecompileHelper;
+use dp_contract::abi_util::abi_encode_bytes;
+// --- paritytech ---
+use fp_evm::{
+	Context, ExitSucceed, Precompile, PrecompileFailure, PrecompileOutput, PrecompileResult,
+};
+
+#[darwinia_evm_precompile_utils::selector]
+enum Action {
+	StateGetStorage = "state_storage(bytes)",
+}
+
+pub trait StorageFilterT {
+	fn allow(prefix: &[u8]) -> bool;
+}
+
+pub struct StateStorage<T, F> {
+	_marker: PhantomData<(T, F)>,
+}
+
+impl<T, F> Precompile for StateStorage<T, F>
+where
+	T: darwinia_evm::Config,
+	F: StorageFilterT,
+{
+	fn execute(
+		input: &[u8],
+		target_gas: Option<u64>,
+		context: &Context,
+		is_static: bool,
+	) -> PrecompileResult {
+		let mut helper = PrecompileHelper::<T>::new(input, target_gas);
+		let (selector, data) = helper.split_input()?;
+		let action = Action::from_u32(selector)?;
+
+		// Check state modifiers
+		helper.check_state_modifier(context, is_static, StateMutability::View)?;
+
+		let output = match action {
+			Action::StateGetStorage => {
+				let tokens = ethabi::decode(&[ParamType::Bytes], data)
+					.map_err(|_| helper.revert("ethabi decoded error"))?;
+				let key = match &tokens[0] {
+					Token::Bytes(bytes) => bytes,
+					_ => return Err(helper.revert("The input param error")),
+				};
+
+				if key.len() < 16 || !F::allow(&key[0..16]) {
```

Comment:
What does this 16 stand for? it's better to define a constant value here.

Reply:
16 = `Twox128::hash(b"EVM")` length. You are right. Give it a constant value is better.